[mlir][Transforms] Support rolling back properties in dialect conversion

matthias-springer opened the pull request:
The dialect conversion rolls back in-place op modifications upon failure. Rolling back modifications of attributes is already supported, but there was no support for properties until now.

joker-eph commented on mlir/lib/Transforms/Utils/DialectConversion.cpp:
```diff
@@ -1002,12 +1002,31 @@ class ModifyOperationRewrite : public OperationRewrite {
+      // Make a copy of the properties.
+      int size = op->getPropertiesStorageSize();
+      propertiesStorage = operator new(size);
+      memcpy(propertiesStorage, prop.as<void *>(), size);
```
That does not seem correct C++ to me: you can't assume that a property can be copied as a "POD" (I don't remember the new terminology).
Properties can be arbitrary C++ objects (think std::string for example).

matthias-springer replied:
Changed to `copyOpProperties`.

joker-eph commented on mlir/lib/Transforms/Utils/DialectConversion.cpp:
```diff
@@ -1002,12 +1002,31 @@ class ModifyOperationRewrite : public OperationRewrite {
+
+  void commit() override {
+    if (propertiesStorage) {
+      operator delete(propertiesStorage);
```
This needs to call the property destructor

joker-eph commented on mlir/lib/Transforms/Utils/DialectConversion.cpp:
```diff
@@ -1016,13 +1035,20 @@ class ModifyOperationRewrite : public OperationRewrite {
+    if (propertiesStorage) {
+      OpaqueProperties prop(propertiesStorage);
+      op->copyProperties(prop);
+      operator delete(propertiesStorage);
```
(Same)

joker-eph commented on mlir/lib/Transforms/Utils/DialectConversion.cpp:
```diff
@@ -1002,12 +1002,31 @@ class ModifyOperationRewrite : public OperationRewrite {
+      // Make a copy of the properties.
+      propertiesStorage = operator new(op->getPropertiesStorageSize());
+      OpaqueProperties propCopy(propertiesStorage);
+      op->getName().copyOpProperties(propCopy, prop);
```
The constructor is never called I believe

matthias-springer replied:
Replaced `copyOpProperties` with `initProperties`, this is calling the copy constructor.

matthias-springer committed:
[mlir][Transforms] Support rolling back properties in dialect conversion

The dialect conversion rolls back inplace op modifications upon failure. Rolling back modifications of op properties was not supported before this commit.
